tools/functional-tester/etcd-runner/main.go

```diff
@@ -212,6 +219,166 @@ func runRacer(eps []string, round int) {
 	doRounds(rcs, round)
 }
 
+func runWatcher(eps []string) {
+	var runningTime time.Duration = 30 // time for which operation should be performed
+	noOfPrefixes := 36                 // total number of prefixes which will be watched upon
+	watchPerPrefix := 10               // number of watchers per prefix
+	reqRate := 20                      // put request per second
+	keyPrePrefix := 100000             // max number of keyPrePrefixs for put operation
+
+	ctx := context.Background()
+	prefixKeysMap := make(map[string][]string) // map for holding prefixes and values
+	// pre generate prefixes equal to noOfPrefixes
+	prefixes := generateUniqueKeys(5, noOfPrefixes)
+	// pre generate keys
+	keys := generateRandomKeys(10, keyPrePrefix)
+
+	var (
+		revision int64
+		wg       sync.WaitGroup
+		pr       *clientv3.PutResponse
+		gr       *clientv3.GetResponse
+		err      error
+	)
+
+	for i := range prefixes {
+		prefix := prefixes[i]
+		// initialize prefix key map for each prefix
+		prefixKeysMap[prefix] = make([]string, 0)
+	}
+
+	// create client for performing get and put operations
+	client := randClient(eps)
+	defer client.Close()
+
+	// get revision using get request
+	gr, err = client.Get(ctx, "non-existant")
+	if err != nil {
+		log.Fatal("Error occured while trying to get the revision.")
+	}
+	revision = gr.Header.Revision
+
+	ctxt, cancel := context.WithDeadline(ctx, time.Now().Add(runningTime*time.Second))
+	defer cancel()
+
+	// generate and put keys in cluster
+	go func() {
+		limiter := rate.NewLimiter(rate.Limit(reqRate), reqRate)
+		prefixIndex := 0
+
+		for i := 0; i < len(keys); i++ {
+			prefix := prefixes[prefixIndex]
+			key := prefix + keys[i]
+			// limit key put as per reqRate
+			err = limiter.Wait(ctxt)
+			if err != nil {
+				break
+			}
+
+			//add all possible updates on prefixes for the generated key
+			for prefixkey, vals := range prefixKeysMap {
+				if strings.HasPrefix(key, prefixkey) {
+					prefixKeysMap[prefixkey] = append(vals, key)
+				}
+			}
+
+			// perform the put operation
+			pr, err = client.Put(ctxt, key, key)
+			if err == context.DeadlineExceeded {
+				break
+			}
+
+			if err != nil {
+				log.Printf("Error: %v occured while trying to key: %v, value : %v to kv store.", err, key, key)
+				continue
+			}
+			prefixIndex++
+			if prefixIndex == len(prefixes) {
+				prefixIndex = 0
+			}
+		}
+	}()
+
+	wg.Add(noOfPrefixes * watchPerPrefix)
+	for i := 0; i < noOfPrefixes; i++ {
+		prefix := prefixes[i]
+		for j := 0; j < watchPerPrefix; j++ {
+			// for all watcher verify the watch out put with actual key updates on prefixes
+			go func(prefix string) {
```

Comment:
we need to make line 307 to 377 a function. the indentation here is a little bit crazy.